TP 1 Terminé

arthurdeschamps opened the pull request:
Les principales fonctionnalités ainsi que quelques détails sont implémentés. Je n'ai détecté aucun bug jusqu'à maintenant, j'espère qu'il n'y en a pas de caché.

Quelques fonctionnalités importantes:
- Gestion des combats avec input de l'utilisateur et gestion automatique de l'ordinateur (par des techniques utilisant l'aléatoire à 100% ou alors en utilisant des "principes" comme  par exemple choisir le pokemon qui a le plus de points de vies pour remplacer un pokemon KO, favoriser l'attaque par rapport aux autres actions , etc
- Laisser l'utilisateur créer des pokemons avant le combat et en générer des aléatoires pour l'ordinateur (même nombre pour les deux)
- Gestion des dégâts infligés correspondant presque entièrement à Bulbapedia (a part les variables liées au temps, au terrain ou aux objets par exemple, soit la variable `other`) 
- Affichage à chaque tour de l'état du combat avec pokemon dans le champ de bataille, en dehors, leurs stats, leur points de vie restants, etc
- Gestion des points d'actions restants pour chaque attaque (diminution des powerpoints, disparition de l'option "attack" lorsqu'il ne reste aucune attaque avec des powerpoints, etc)
- Possibilité de s'enfuir pour l'utilisateur comme pour l'ordinateur suivant une formule probabiliste
- Possibilité de switcher deux Pokemon


Fonctionnalités manquantes:
- Gestion des items (l'action "item" proposée retournera toujours que le sac ne contient aucun objet, que ce soit du côté utilisateur ou du côté ordinateur)
- Impact du temps sur le combat
- Impact du terrain sur le combat

Commit:
Todos finished: some moves created, some species created, effective_stats calculation, typeModifier function, damage (two remaining things to implement : the parameter “other” and the “stages” for critical hits probabilities) No bug/logic testing yet, so there might be a lot of code to correct Big todos remaining: battle state and battle

Commit:
Added initialisation call in the tests file in order to simulate the battle (and track any potential bug or inconsistencie).

swift_exercices file greatly improved and probably working without bugs.

User’s turn is done, but computer’s turn and actions resolving still need to be done.

Commit:
Battles work until someone loses with two possible actions: attacking or running away.
Still very important things to implement: Switch pokemons and use items (not sure if that will be ever implemented).
Powerpoints stay constant for now: that’s the first step for future dev.

Commit:
Moved in init function moves and species definitions.
Switch pokemons feature added.

Commit:
Added possibility to choose a custom Pokemon based on an existing species.
Current bug: computer doesn’t make any action anymore

Commit:
Fix: added missing cases during actions resolving (which causes the previous bug with the opponent not doing his action)
Main features are now implemented.

Commit:
Done:
- Nicknames are now randomly chosen for computer’s pokemons (until the number of pokemons reaches the total amount of available names)
- Changed type of user_action from Any? To Move?, since it’s wether a move or a nil when the action is resolved immediately
- Computer choses the Pokemon with the greatest hitpoints remaining when it comes to make a substitution (or when a Pokemon is dazed)
- Random natures are generated for newly created Pokemons (user or computer)

Commit:
Merge branch 'dev'

* dev:
  Done: - Nicknames are now randomly chosen for computer’s pokemons (until the number of pokemons reaches the total amount of available names) - Changed type of user_action from Any? To Move?, since it’s wether a move or a nil when the action is resolved immediately - Computer choses the Pokemon with the greatest hitpoints remaining when it comes to make a substitution (or when a Pokemon is dazed) - Random natures are generated for newly created Pokemons (user or computer)
  Fix: added missing cases during actions resolving (which causes the previous bug with the opponent not doing his action) Main features are now implemented.
  Added possibility to choose a custom Pokemon based on an existing species. Current bug: computer doesn’t make any action anymore
  Moved in init function moves and species definitions. Switch pokemons feature added.

saucisson commented:
Your code does not compile on the continuous integration platform, because it requires the `Darwin` package, that does not exist on Linux. What functions do you use from this package?

arthurdeschamps commented:
I'm using arc4_random and exit that requires Darwin. I'll try to get rid of them and find a replacement.

arthurdeschamps committed:
Darwin not required anymore

Functions from Darwin (exit and arc4random_uniform) are now defined  for linux platform

arthurdeschamps committed:
Merge branch 'dev'
